                                                      MovieLens_Casestudy

In [1]:
pip install numpy

Note: you may need to restart the kernel to use updated packages.



[notice] A new release of pip is available: 25.0.1 -> 25.2
[notice] To update, run: python.exe -m pip install --upgrade pip


In [2]:
pip install pandas

Note: you may need to restart the kernel to use updated packages.



[notice] A new release of pip is available: 25.0.1 -> 25.2
[notice] To update, run: python.exe -m pip install --upgrade pip


In [3]:
#Let us start by importing the required libraries 
import pandas as pd
import numpy as np

                                                       Reading the Datasets
                                                         1.User dataset

In [5]:
# Reading datasets by using read_csv from pandas package
movie = pd.read_csv("E:/Data Analytics with python/Case studies/movie.csv")
user = pd.read_csv("E:/Data Analytics with python/Case studies/user.csv")
ratings = pd.read_csv("E:/Data Analytics with python/Case studies/ratings.csv")

In [6]:
user.head()   #Displaying the first 5 rows of the user dataset

user id  age gender  occupation zip code
0        1   24      M  technician    85711
1        2   53      F       other    94043
2        3   23      M      writer    32067
3        4   24      M  technician    43537
4        5   33      F       other    15213

In [7]:
user.shape #Understanding the shape of the user dataset

(943, 5)

In [8]:
user.info() #Checking the datatypes of the columns for the user dataset

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 943 entries, 0 to 942
Data columns (total 5 columns):
 #   Column      Non-Null Count  Dtype 
---  ------      --------------  ----- 
 0   user id     943 non-null    int64 
 1   age         943 non-null    int64 
 2   gender      943 non-null    object
 3   occupation  943 non-null    object
 4   zip code    943 non-null    object
dtypes: int64(2), object(3)
memory usage: 37.0+ KB


In [10]:
user.describe()  #Getting the statistical summary of the user dataset

user id         age
count  943.000000  943.000000
mean   472.000000   34.051962
std    272.364951   12.192740
min      1.000000    7.000000
25%    236.500000   25.000000
50%    472.000000   31.000000
75%    707.500000   43.000000
max    943.000000   73.000000

                                                            2.Movie Dataset

In [11]:
movie.head() #Displaying the first five rows of the movie dataset

movie id  movie title release date  Action  Adventure  Animation  \
0         1   Toy Story      1-Jan-95       0          0          1   
1         2   GoldenEye      1-Jan-95       1          1          0   
2         3  Four Rooms      1-Jan-95       0          0          0   
3         4  Get Shorty      1-Jan-95       1          0          0   
4         5     Copycat      1-Jan-95       0          0          0   

   Childrens  Comedy  Crime  Documentary  ...  Fantasy  Film-Noir  Horror  \
0          1       1      0            0  ...        0          0       0   
1          0       0      0            0  ...        0          0       0   
2          0       0      0            0  ...        0          0       0   
3          0       1      0            0  ...        0          0       0   
4          0       0      1            0  ...        0          0       0   

   Musical  Mystery  Romance  Sci-Fi  Thriller  War  Western  
0        0        0        0       0         0    0        0  
1        0        0        0       0         1    0        0  
2        0        0        0       0         1    0        0  
3        0        0        0       0         0    0        0  
4        0        0        0       0         1    0        0  

[5 rows x 21 columns]

In [13]:
movie.shape #Understanding the shape of movie dataset

(1680, 21)

In [14]:
movie.info()  #Checking the datatypes of the columns for the movie dataset

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1680 entries, 0 to 1679
Data columns (total 21 columns):
 #   Column        Non-Null Count  Dtype 
---  ------        --------------  ----- 
 0   movie id      1680 non-null   int64 
 1   movie title   1680 non-null   object
 2   release date  1680 non-null   object
 3   Action        1680 non-null   int64 
 4   Adventure     1680 non-null   int64 
 5   Animation     1680 non-null   int64 
 6   Childrens     1680 non-null   int64 
 7   Comedy        1680 non-null   int64 
 8   Crime         1680 non-null   int64 
 9   Documentary   1680 non-null   int64 
 10  Drama         1680 non-null   int64 
 11  Fantasy       1680 non-null   int64 
 12  Film-Noir     1680 non-null   int64 
 13  Horror        1680 non-null   int64 
 14  Musical       1680 non-null   int64 
 15  Mystery       1680 non-null   int64 
 16  Romance       1680 non-null   int64 
 17  Sci-Fi        1680 non-null   int64 
 18  Thriller      1680 non-null   int64 
 19  War   

In [15]:
movie.describe() #Getting the statistical summary for the movie dataset

movie id       Action    Adventure    Animation    Childrens  \
count  1680.000000  1680.000000  1680.000000  1680.000000  1680.000000   
mean    841.525595     0.149405     0.080357     0.025000     0.072619   
std     485.609591     0.356593     0.271926     0.156171     0.259587   
min       1.000000     0.000000     0.000000     0.000000     0.000000   
25%     421.750000     0.000000     0.000000     0.000000     0.000000   
50%     841.500000     0.000000     0.000000     0.000000     0.000000   
75%    1261.250000     0.000000     0.000000     0.000000     0.000000   
max    1682.000000     1.000000     1.000000     1.000000     1.000000   

            Comedy        Crime  Documentary        Drama      Fantasy  \
count  1680.000000  1680.000000  1680.000000  1680.000000  1680.000000   
mean      0.300595     0.064881     0.029762     0.431548     0.013095   
std       0.458653     0.246389     0.169980     0.495440     0.113717   
min       0.000000     0.000000     0.000000     0.000000     0.000000   
25%       0.000000     0.000000     0.000000     0.000000     0.000000   
50%       0.000000     0.000000     0.000000     0.000000     0.000000   
75%       1.000000     0.000000     0.000000     1.000000     0.000000   
max       1.000000     1.000000     1.000000     1.000000     1.000000   

         Film-Noir       Horror      Musical      Mystery      Romance  \
count  1680.000000  1680.000000  1680.000000  1680.000000  1680.000000   
mean      0.014286     0.054762     0.033333     0.036310     0.147024   
std       0.118701     0.227583     0.179559     0.187115     0.354235   
min       0.000000     0.000000     0.000000     0.000000     0.000000   
25%       0.000000     0.000000     0.000000     0.000000     0.000000   
50%       0.000000     0.000000     0.000000     0.000000     0.000000   
75%       0.000000     0.000000     0.000000     0.000000     0.000000   
max       1.000000     1.000000     1.000000     1.000000     1.000000   

            Sci-Fi     Thriller          War      Western  
count  1680.000000  1680.000000  1680.000000  1680.000000  
mean      0.060119     0.149405     0.042262     0.016071  
std       0.237778     0.356593     0.201246     0.125788  
min       0.000000     0.000000     0.000000     0.000000  
25%       0.000000     0.000000     0.000000     0.000000  
50%       0.000000     0.000000     0.000000     0.000000  
75%       0.000000     0.000000     0.000000     0.000000  
max       1.000000     1.000000     1.000000     1.000000

                                                           3.Markdown

In [17]:
ratings.head() #Displaying the first five rows of the ratings dataset

user id  movie id  rating  timestamp
0      196       242       3  881250949
1      186       302       3  891717742
2       22       377       1  878887116
3      244        51       2  880606923
4      166       346       1  886397596

In [19]:
ratings.shape #Understanding the shape of ratings dataset

(100000, 4)

In [20]:
ratings.info() #Checking the datatypes of the columns for the ratings dataset

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 100000 entries, 0 to 99999
Data columns (total 4 columns):
 #   Column     Non-Null Count   Dtype
---  ------     --------------   -----
 0   user id    100000 non-null  int64
 1   movie id   100000 non-null  int64
 2   rating     100000 non-null  int64
 3   timestamp  100000 non-null  int64
dtypes: int64(4)
memory usage: 3.1 MB


In [21]:
ratings.describe() #Getting the statistical summary for the movie dataset

user id       movie id         rating     timestamp
count  100000.00000  100000.000000  100000.000000  1.000000e+05
mean      462.48475     425.530130       3.529860  8.835289e+08
std       266.61442     330.798356       1.125674  5.343856e+06
min         1.00000       1.000000       1.000000  8.747247e+08
25%       254.00000     175.000000       3.000000  8.794487e+08
50%       447.00000     322.000000       4.000000  8.828269e+08
75%       682.00000     631.000000       4.000000  8.882600e+08
max       943.00000    1682.000000       5.000000  8.932866e+08

                 #Let's extract some meaningful insights from the data by answering a few-business oriented questions
                                             Q1.How many movies belongs to a particular genre?

In [24]:
#Let's analyze using the pandas functions
movie.head()

movie id  movie title release date  Action  Adventure  Animation  \
0         1   Toy Story      1-Jan-95       0          0          1   
1         2   GoldenEye      1-Jan-95       1          1          0   
2         3  Four Rooms      1-Jan-95       0          0          0   
3         4  Get Shorty      1-Jan-95       1          0          0   
4         5     Copycat      1-Jan-95       0          0          0   

   Childrens  Comedy  Crime  Documentary  ...  Fantasy  Film-Noir  Horror  \
0          1       1      0            0  ...        0          0       0   
1          0       0      0            0  ...        0          0       0   
2          0       0      0            0  ...        0          0       0   
3          0       1      0            0  ...        0          0       0   
4          0       0      1            0  ...        0          0       0   

   Musical  Mystery  Romance  Sci-Fi  Thriller  War  Western  
0        0        0        0       0         0    0        0  
1        0        0        0       0         1    0        0  
2        0        0        0       0         1    0        0  
3        0        0        0       0         0    0        0  
4        0        0        0       0         1    0        0  

[5 rows x 21 columns]

In [27]:
movie.columns  #Getting all the columns of the movie dataset

Index(['movie id', 'movie title', 'release date', 'Action', 'Adventure',
       'Animation', 'Childrens', 'Comedy', 'Crime', 'Documentary', 'Drama',
       'Fantasy', 'Film-Noir', 'Horror', 'Musical', 'Mystery', 'Romance',
       'Sci-Fi', 'Thriller', 'War', 'Western'],
      dtype='object')

In [32]:
#Since,there are many genres,let's create a variable to store all the values as a list so that we can use it for further analysis
genres = [ 'Action', 'Adventure','Animation',
           'Childrens', 'Comedy', 'Crime', 'Documentary', 'Drama',
       'Fantasy', 'Film-Noir','Horror', 'Musical', 'Mystery', 
       'Romance','Sci-Fi', 'Thriller', 'War', 'Western']

In [34]:
#Using the length function to calculate the number of genres
print("The number of genres present in the movie dataset are",len(genres))

The number of genres present in the movie dataset are 18


In [36]:
#To find the answer-here, we can use the genre list instead of taking each column and adding it
genres #printing the genres list

['Action',
 'Adventure',
 'Animation',
 'Childrens',
 'Comedy',
 'Crime',
 'Documentary',
 'Drama',
 'Fantasy',
 'Film-Noir',
 'Horror',
 'Musical',
 'Mystery',
 'Romance',
 'Sci-Fi',
 'Thriller',
 'War',
 'Western']

In [37]:
movie.head()

movie id  movie title release date  Action  Adventure  Animation  \
0         1   Toy Story      1-Jan-95       0          0          1   
1         2   GoldenEye      1-Jan-95       1          1          0   
2         3  Four Rooms      1-Jan-95       0          0          0   
3         4  Get Shorty      1-Jan-95       1          0          0   
4         5     Copycat      1-Jan-95       0          0          0   

   Childrens  Comedy  Crime  Documentary  ...  Fantasy  Film-Noir  Horror  \
0          1       1      0            0  ...        0          0       0   
1          0       0      0            0  ...        0          0       0   
2          0       0      0            0  ...        0          0       0   
3          0       1      0            0  ...        0          0       0   
4          0       0      1            0  ...        0          0       0   

   Musical  Mystery  Romance  Sci-Fi  Thriller  War  Western  
0        0        0        0       0         0    0        0  
1        0        0        0       0         1    0        0  
2        0        0        0       0         1    0        0  
3        0        0        0       0         0    0        0  
4        0        0        0       0         1    0        0  

[5 rows x 21 columns]

In [40]:
movie[genres].sum() 

Action         251
Adventure      135
Animation       42
Childrens      122
Comedy         505
Crime          109
Documentary     50
Drama          725
Fantasy         22
Film-Noir       24
Horror          92
Musical         56
Mystery         61
Romance        247
Sci-Fi         101
Thriller       251
War             71
Western         27
dtype: int64

In [41]:
movie.head()

movie id  movie title release date  Action  Adventure  Animation  \
0         1   Toy Story      1-Jan-95       0          0          1   
1         2   GoldenEye      1-Jan-95       1          1          0   
2         3  Four Rooms      1-Jan-95       0          0          0   
3         4  Get Shorty      1-Jan-95       1          0          0   
4         5     Copycat      1-Jan-95       0          0          0   

   Childrens  Comedy  Crime  Documentary  ...  Fantasy  Film-Noir  Horror  \
0          1       1      0            0  ...        0          0       0   
1          0       0      0            0  ...        0          0       0   
2          0       0      0            0  ...        0          0       0   
3          0       1      0            0  ...        0          0       0   
4          0       0      1            0  ...        0          0       0   

   Musical  Mystery  Romance  Sci-Fi  Thriller  War  Western  
0        0        0        0       0         0    0        0  
1        0        0        0       0         1    0        0  
2        0        0        0       0         1    0        0  
3        0        0        0       0         0    0        0  
4        0        0        0       0         1    0        0  

[5 rows x 21 columns]

In [42]:
#Using Pandas sort_values() Functions
movie[genres].sum().sort_values(ascending = False) #Now,the values will be sorted in descending order

Drama          725
Comedy         505
Thriller       251
Action         251
Romance        247
Adventure      135
Childrens      122
Crime          109
Sci-Fi         101
Horror          92
War             71
Mystery         61
Musical         56
Documentary     50
Animation       42
Western         27
Film-Noir       24
Fantasy         22
dtype: int64

INSIGHTS:
    Movies are distributed across 18 genres in total
    We can observe that the 'Drama' genre has the highest  number of movies total followed by the 'Comedy'
    The top five genres with the highest number of movies are 'Drama','Comedy','Action','Thriller' and 'Romance'
    'Fantasy' genre has the lowest number of movies

                                     Q2.Can we find the movies that have more than one genre?

In [43]:
#To find the information let us create a new column named Number of Genres and add it to the dataframe
movie["Number of Genres"]=movie.loc[:,genres].sum(axis=1) #it's a list now(num of genres)
#the new column adds the summation of one's from the genre columns along the row of a particular movie
#movie.loc[:,genre]-Picking up all the rows and columns that are present in the genres list
#axis=1--this parameter is used to get the summation of values along the columns

In [49]:
movie[["movie id","Number of Genres"]]

movie id  Number of Genres
0            1                 3
1            2                 3
2            3                 1
3            4                 3
4            5                 3
...        ...               ...
1675      1678                 1
1676      1679                 2
1677      1680                 2
1678      1681                 1
1679      1682                 1

[1680 rows x 2 columns]

In [44]:
#we can print the head to see whether the new column has been added or not 
movie.head()

movie id  movie title release date  Action  Adventure  Animation  \
0         1   Toy Story      1-Jan-95       0          0          1   
1         2   GoldenEye      1-Jan-95       1          1          0   
2         3  Four Rooms      1-Jan-95       0          0          0   
3         4  Get Shorty      1-Jan-95       1          0          0   
4         5     Copycat      1-Jan-95       0          0          0   

   Childrens  Comedy  Crime  Documentary  ...  Film-Noir  Horror  Musical  \
0          1       1      0            0  ...          0       0        0   
1          0       0      0            0  ...          0       0        0   
2          0       0      0            0  ...          0       0        0   
3          0       1      0            0  ...          0       0        0   
4          0       0      1            0  ...          0       0        0   

   Mystery  Romance  Sci-Fi  Thriller  War  Western  Number of Genres  
0        0        0       0         0    0        0                 3  
1        0        0       0         1    0        0                 3  
2        0        0       0         1    0        0                 1  
3        0        0       0         0    0        0                 3  
4        0        0       0         1    0        0                 3  

[5 rows x 22 columns]

In [52]:
#Let us pick up the rows where number of Genres>1 so we can easily find the num. of movies with >1
movie[movie['Number of Genres']>1] #this will subset the data as per the condition

movie id         movie title release date  Action  Adventure  Animation  \
0            1          Toy Story      1-Jan-95       0          0          1   
1            2          GoldenEye      1-Jan-95       1          1          0   
3            4         Get Shorty      1-Jan-95       1          0          0   
4            5            Copycat      1-Jan-95       0          0          0   
6            7     Twelve Monkeys      1-Jan-95       0          0          0   
...        ...                 ...          ...     ...        ...        ...   
1666      1669  MURDER and murder     20-Jun-97       0          0          0   
1667      1670            Tainted      1-Feb-98       0          0          0   
1670      1673             Mirage      1-Jan-95       1          0          0   
1676      1679          B. Monkey      6-Feb-98       0          0          0   
1677      1680      Sliding Doors      1-Jan-98       0          0          0   

      Childrens  Comedy  Crime  Documentary  ...  Film-Noir  Horror  Musical  \
0             1       1      0            0  ...          0       0        0   
1             0       0      0            0  ...          0       0        0   
3             0       1      0            0  ...          0       0        0   
4             0       0      1            0  ...          0       0        0   
6             0       0      0            0  ...          0       0        0   
...         ...     ...    ...          ...  ...        ...     ...      ...   
1666          0       0      1            0  ...          0       0        0   
1667          0       1      0            0  ...          0       0        0   
1670          0       0      0            0  ...          0       0        0   
1676          0       0      0            0  ...          0       0        0   
1677          0       0      0            0  ...          0       0        0   

      Mystery  Romance  Sci-Fi  Thriller  War  Western  Number of Genres  
0           0        0       0         0    0        0                 3  
1           0        0       0         1    0        0                 3  
3           0        0       0         0    0        0                 3  
4           0        0       0         1    0        0                 3  
6           0        0       1         0    0        0                 2  
...       ...      ...     ...       ...  ...      ...               ...  
1666        1        0       0         0    0        0                 3  
1667        0        0       0         1    0        0                 2  
1670        0        0       0         1    0        0                 2  
1676        0        1       0         1    0        0                 2  
1677        0        1       0         0    0        0                 2  

[849 rows x 22 columns]

In [53]:
#Let us print the shape of the dataframe so that we know the number of movies that are having more than 1 genre 
movie[movie['Number of Genres']>1].shape   #Number of genres contain 849 dataset

(849, 22)

In [56]:
movie.shape #original movie dataset

(1680, 22)

                         Q3.Can we find the which movie has been  most preferred by the users?                                          

In [59]:
#We can merge both the datasets on the basis of common variables that is movie_id
df_merge = movie.merge(ratings,on = 'movie id',how = 'inner')
df_merge.head()

movie id movie title release date  Action  Adventure  Animation  Childrens  \
0         1  Toy Story      1-Jan-95       0          0          1          1   
1         1  Toy Story      1-Jan-95       0          0          1          1   
2         1  Toy Story      1-Jan-95       0          0          1          1   
3         1  Toy Story      1-Jan-95       0          0          1          1   
4         1  Toy Story      1-Jan-95       0          0          1          1   

   Comedy  Crime  Documentary  ...  Mystery  Romance  Sci-Fi  Thriller  War  \
0       1      0            0  ...        0        0       0         0    0   
1       1      0            0  ...        0        0       0         0    0   
2       1      0            0  ...        0        0       0         0    0   
3       1      0            0  ...        0        0       0         0    0   
4       1      0            0  ...        0        0       0         0    0   

   Western  Number of Genres  user id  rating  timestamp  
0        0                 3      308       4  887736532  
1        0                 3      287       5  875334088  
2        0                 3      148       4  877019411  
3        0                 3      280       4  891700426  
4        0                 3       66       3  883601324  

[5 rows x 25 columns]

In [60]:
genres

['Action',
 'Adventure',
 'Animation',
 'Childrens',
 'Comedy',
 'Crime',
 'Documentary',
 'Drama',
 'Fantasy',
 'Film-Noir',
 'Horror',
 'Musical',
 'Mystery',
 'Romance',
 'Sci-Fi',
 'Thriller',
 'War',
 'Western']

In [65]:
#Running the loop across the merged dataframes for the columns that are present in the genres list
for i in genres:
    print(i,':' ,df_merge[df_merge[i]==1].rating.mean())

Action : 3.480245417953027
Adventure : 3.503526503308369
Animation : 3.5766990291262135
Childrens : 3.3532442216652742
Comedy : 3.3940734781442745
Crime : 3.6322780881440098
Documentary : 3.6728232189973613
Drama : 3.6873793708484772
Fantasy : 3.2152366863905324
Film-Noir : 3.9215233698788228
Horror : 3.2903893172841827
Musical : 3.521396851029471
Mystery : 3.63813155386082
Romance : 3.621704948358255
Sci-Fi : 3.5607227022780834
Thriller : 3.5090069495245064
War : 3.815811874866993
Western : 3.6132686084142396


INSIGHTS:
    'Film-Noir' genre is having the highest average rating followed by the 'War' genre
    'Fantasy' genre is having the lowest average ratings
    Among 18 genres only 5 genres has been received a rating below 3.5 which means the 72% of genres are rated above average